web/book/src/internals/modules.md

```diff
@@ -0,0 +1,311 @@
+# Modules
+
+> Status: proposal
+
+> This is a technical document. For a "how to use" or a TLDR; skip to the
+> [Example](#Example) section.
+
+Design goals:
+
+1. Allow importing declarations from other files.
+
+2. Have namespaces for things like `std`.
+
+3. Have hierarchical structure so we can represent files in directories.
+
+4. Have unambiguous module structure within a project.
+
+## Definition
+
+Module is a namespace that contains declarations. A module declaration itself,
+which means that it can contain nested child modules.
+
+This means that modules form a
+[tree graph](<https://en.wikipedia.org/wiki/Tree_(graph_theory)>), which we call
+"the module structure".
+
+For the sake of this document, we will express the module structure with
+`module` keyword and a code block encased in curly braces:
+
+```
+module my_playlists {
+    let bangers = ... # a declaration
+
+    module soundtracks {
+        let movie_albums = ... # another declaration
+    }
+}
+```
+
+> The syntax `module name { ...decls... }` is not part of PRQL language, with
+> the objection that it is unnecessary as it only adds more ways of defining
+> modules. If a significant upside of this syntax is found, it may be added in
+> the future.
+
+## Name resolution
+
+Any declarations within a module can be referenced from the outside of the
+module:
+
+```prql_no_test
+# using module structure declared above
+
+let great_tracks = my_playlists.bangers
+
+let movie_scores = my_playlists.soundtracks.movie_albums
+```
+
+Identifiers are resolved relative to current module.
+
+```prql_no_test
+module my_playlists {
+    module soundtracks {
+        let movie_albums = (from albums | filter id == 3)
+    }
+
+    from soundtracks.movie_albums
+}
+from my_playlists.soundtracks.movie_albums
+```
+
+If an identifier cannot be resolved relative to the current module, it tries to
+resolve relative to the parent module. This is repeated, stepping up the module
+hierarchy until a match is found or root of the module structure is reached.
+
+```prql_no_test
+module my_playlists {
+    let decl_1 = ...
+
+    module soundtracks {
+        let decl_2 = ...
+    }
+
+    module upbeat_rock {
+        let decl_3 = ...
+
+        from decl_1 | join soundtracks.decl2 | join decl_3
+    }
+}
+```
```

Comment:
I'm not sure about this.

Upsides:

- the std module can be defined just in the root module and be accessible
  everywhere under name `std`,

Downsides:

- we need to find the root of the module structure,

Reply:
Is each module required to be imported? If so, is there an implicit `module std`?

Reply:
There's some possible modules-as-inheritance approaches from the implicit shadowing.

But also possible some quite confusing error messages when a parent changes or adds a definition!

Overall, no strong view. If no one is confident, possibly better to start with the explicit approach and then we can add the implicit one on later?

Reply:
> When does this know when to stop going up the file tree?

Well, at the root? So `std` can be defined in the root only.

> But also possible some quite confusing error messages when a parent changes or adds a definition!

If we give declarations in child priority over declarations in parent, no change in parent will be able to overshadow the child declaration.

But it could happen that a module would be referring to a grand-parent module and a change in parent would break the child.

```
 grandparent -> decl
      parent -> we add decl here
      child  -> ref to decl now resolves to parent instead of grandparent
```

Reply:
> > When does this know when to stop going up the file tree?
> 
> Well, at the root? So `std` can be defined in the root only.

This is moot if we have a file that can only be at the root, such as `_project.prql` or `project.prql` — we can stop there

But otherwise we need some way of knowing when to stop searching upwards, so we don't keep going to `/`. We could do that when we no longer find a match of `$dir` containing `_$dir.prql`...

I also think this "keep searching up" would be fine to push into a later version if it's not trivial to implement — very reasonable to adjust at implementation time. Though we'd need a way to reference a parent module, such as `module ..foo`.

Reply:
Oh, this is a tad different thing I was talking about here.

Here I meant name resolution only, which happens after the module structure has been read from files, parsed and resolved.

When deciding what files to load, the compiler should get the path to the root module and then find references to other files which need to be loaded from there.